pkgs/applications/virtualization/qemu/encode-video.c

```diff
@@ -0,0 +1,737 @@
+#include <fcntl.h>
+#include <stdbool.h>
+#include <byteswap.h>
+
+#include <libavformat/avformat.h>
+#include <libswscale/swscale.h>
+
+#include <zlib.h>
+
+/* This is needed in order to distinguish between real errors and read failures
+ * indicating EOF, so that we can exit gracefully if errors happen.
+ */
+static bool is_error;
+
+struct bound_info {
+    uint32_t width;
+    uint32_t height;
+    uint64_t start_time;
+    size_t frames;
+};
+
+struct packet_switch {
+    uint32_t width;
+    uint32_t height;
+    enum AVPixelFormat format;
+    uint8_t bpp;
+};
+
+struct packet_update {
+    uint32_t x;
+    uint32_t y;
+    uint32_t w;
+    uint32_t h;
+    uint64_t timestamp;
+    size_t datalen;
+};
+
+/* Use AVIO to open the file if the URL is pointing to a file instead of for
+ * example streaming it over network.
+ */
+static bool maybe_open_avio(AVFormatContext *context, const char *uri)
+{
+    int ret;
+
+    if (context->oformat->flags & AVFMT_NOFILE)
+        return true;
+
+    if ((ret = avio_open(&context->pb, uri, AVIO_FLAG_WRITE)) < 0) {
+        fprintf(stderr, "Unable to open '%s': %s\n", uri, av_err2str(ret));
+        return false;
+    }
+
+    return true;
+}
+
+/* Close AVIO if the passed URL is a file. */
+static void maybe_close_avio(AVFormatContext *context)
+{
+    if (context->oformat->flags & AVFMT_NOFILE)
+        return;
+
+    avio_closep(&context->pb);
+}
+
+/* Read bytes from the input stream with the given length and allocate a buffer
+ * large enough.
+ */
+static void *alloc_read(gzFile ifile, size_t len)
+{
+    void *buf;
+    int i, readlen;
+
+    is_error = false;
+
+    if ((buf = malloc(len)) == NULL) {
+        fprintf(stderr, "Unable to allocate buffer for "
+                "intermediate video packet: %s\n", strerror(errno));
+        is_error = true;
+        return NULL;
+    }
+
+    for (i = 0; i < len; i += readlen) {
+        readlen = gzread(ifile, buf + i, len - i);
+        if (readlen <= 0) {
+            free(buf);
+            /* Don't print an error because we want to make sure that whenever
+             * the stream ends prematurely, we still do have a valid video.
+             */
+            if (readlen < 0) {
+                fprintf(stderr, "Unable to read %zu bytes from input file.",
+                        len - i);
+                is_error = true;
+            }
+            return NULL;
+        }
+    }
+
+    return buf;
+}
+
+/* The packed pixel format from QEMU (and thus Pixman) is using 2 or 4 byte
+ * sequences which are in native host byte order. FFmpeg on the other hand
+ * expects 8-bit sequences, so we need to convert it to big endian as all of
+ * the pixel formats we've choosen for FFmpeg are big endian (either explicitly
+ * or implicitly).
+ */
+static bool convert_endian(void *data, size_t len, uint8_t bpp)
+{
+    const uint16_t endian_test = 1;
+
+    if (*(uint8_t*)&endian_test == 0)
+        return true;
```

Comment:
couldn't we simply use ntoh/hton functions from arpa/inet.h to have functions that simply do the right thing on the right hosts platforms?

Reply:
The reason why I didn't use `htons`/`htonl` was because it would recurse the whole frame even if we didn't need to convert anything.